protocols/relay/src/v2/protocol/outbound_hop.rs

```diff
@@ -0,0 +1,221 @@
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::v2::message_proto::{hop_message, HopMessage, Peer, Status};
+use crate::v2::protocol::{HOP_PROTOCOL_NAME, MAX_MESSAGE_SIZE};
+use asynchronous_codec::{Framed, FramedParts};
+use bytes::Bytes;
+use futures::{future::BoxFuture, prelude::*};
+use futures_timer::Delay;
+use libp2p_core::{upgrade, Multiaddr, PeerId};
+use libp2p_swarm::NegotiatedSubstream;
+use prost::Message;
+use std::convert::TryFrom;
+use std::io::Cursor;
+use std::iter;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use thiserror::Error;
+use unsigned_varint::codec::UviBytes;
+
+pub enum Upgrade {
+    Reserve,
+    Connect { dst_peer_id: PeerId },
+}
+
+impl upgrade::UpgradeInfo for Upgrade {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(HOP_PROTOCOL_NAME)
+    }
+}
+
+impl upgrade::OutboundUpgrade<NegotiatedSubstream> for Upgrade {
+    type Output = Output;
+    type Error = UpgradeError;
+    type Future = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn upgrade_outbound(self, substream: NegotiatedSubstream, _: Self::Info) -> Self::Future {
+        let msg = match self {
+            Upgrade::Reserve => HopMessage {
+                r#type: hop_message::Type::Reserve.into(),
+                peer: None,
+                reservation: None,
+                limit: None,
+                status: None,
+            },
+            Upgrade::Connect { dst_peer_id } => HopMessage {
+                r#type: hop_message::Type::Connect.into(),
+                peer: Some(Peer {
+                    id: dst_peer_id.to_bytes(),
+                    addrs: vec![],
+                }),
+                reservation: None,
+                limit: None,
+                status: None,
+            },
+        };
+
+        let mut encoded_msg = Vec::new();
+        msg.encode(&mut encoded_msg)
+            .expect("Vec to have sufficient capacity.");
+
+        let mut codec = UviBytes::default();
+        codec.set_max_len(MAX_MESSAGE_SIZE);
+        let mut substream = Framed::new(substream, codec);
+
+        async move {
+            substream.send(std::io::Cursor::new(encoded_msg)).await?;
+            let msg: bytes::BytesMut = substream
+                .next()
+                .await
+                .ok_or_else(|| std::io::Error::new(std::io::ErrorKind::UnexpectedEof, ""))??;
+
+            let HopMessage {
+                r#type,
+                peer: _,
+                reservation,
+                limit: _,
```

Comment:
:+1: Exposed via `NetworkBehaviour::OutEvent` for outgoing reservations and outgoing / incoming circuits. See 45f7e218203cc17d132d253b7dad615737507a12.

>  Apart from that, doesn't limit.duration also may be useful for the client behaviour itself. If the relay disconnects our connection to the src because of this, we could directly attempt to re-establish the connection?

I would prefer implementing re-connects along with `libp2p-autorelay`. See https://github.com/libp2p/rust-libp2p/pull/2059#discussion_r771514669. Hope that works for you?

Reply:
Looks good to me.